### PR TITLE
Port Hibernate's lazy attribute fetch groups

### PR DESCRIPTION
**Pull request by maca88**
This PR ports:
[HHH-10267](https://github.com/hibernate/hibernate-orm/commit/1e44e7420b7af10adc9547a461f02a6d979624f1) - Support defining lazy attribute fetch groups
[HHH-11506](https://github.com/hibernate/hibernate-orm/commit/aaebcdf454d57a640c8194d0559642d405f00868) - Fix update of lazy attributes when there are attributes not loaded

Additionally, this PR updates the cache after a lazy group is initialized to be consistent with #1922.

Should be reviewed after #1922.

### CODE REVIEW

**Comment by hazzik** [comment]
@maca88 can you please rebase?

**Comment by maca88** [comment]
Rebased.

**Comment by maca88** [comment]
Rebased.

File: src/NHibernate/Bytecode/IBytecodeEnhancementMetadata.cs [hunk 1]
```diff
@@ -40,11 +40,9 @@ public interface IBytecodeEnhancementMetadata
 		/// Build and inject an interceptor instance into the enhanced entity.
 		/// </summary>
 		/// <param name="entity">The entity into which built interceptor should be injected.</param>
-		/// <param name="lazyPropertiesAreUnfetched">Whether all lazy properties were unfetched or not.</param>
 		/// <param name="session">The session to which the entity instance belongs.</param>
 		/// <returns>The built and injected interceptor.</returns>
-		// TODO: Remove lazyPropertiesAreUnfetched when interceptor will be injected on entity instantiation
-		IFieldInterceptor InjectInterceptor(object entity, bool lazyPropertiesAreUnfetched, ISessionImplementor session);
+		IFieldInterceptor InjectInterceptor(object entity, ISessionImplementor session);
```
Comment (maca88) [inline src/NHibernate/Bytecode/IBytecodeEnhancementMetadata.cs]: The interface is not yet released.

File: src/NHibernate/Bytecode/IBytecodeEnhancementMetadata.cs [hunk 2]
```diff
@@ -66,5 +64,12 @@ public interface IBytecodeEnhancementMetadata
 		/// <param name="entityState">The entity state from which to retrieve the uninitialized lazy properties.</param>
 		/// <returns>The uninitialized property names.</returns>
 		ISet<string> GetUninitializedLazyProperties(object[] entityState);
+
+		/// <summary>
+		/// Check whether the enhanced entity has any uninitialized lazy properties.
+		/// </summary>
+		/// <param name="entity">The entity to check for uninitialized lazy properties.</param>
+		/// <returns>Whether the enhanced entity has any uninitialized lazy properties.</returns>
+		bool HasAnyUninitializedLazyProperties(object entity);
```
Comment (maca88) [inline src/NHibernate/Bytecode/IBytecodeEnhancementMetadata.cs]: The interface is not yet released.

**Comment by fredericDelaporte** [comment]
The documentation should be updated. ([Here](https://nhibernate.info/doc/nhibernate-reference/mapping.html#mapping-declaration-property) and [here](https://nhibernate.info/doc/nhibernate-reference/mapping.html#mapping-declaration-component), so in sources, [here](https://github.com/nhibernate/nhibernate-core/blob/f0f2a53fb778daa06dc7463625ff555ea6fb51ce/doc/reference/modules/basic_mapping.xml#L1668) and [here](https://github.com/nhibernate/nhibernate-core/blob/f0f2a53fb778daa06dc7463625ff555ea6fb51ce/doc/reference/modules/basic_mapping.xml#L2581).)

File: src/NHibernate/Mapping/ByCode/IComponentMapper.cs [hunk 3]
```diff
@@ -31,4 +33,29 @@ public interface IComponentAttributesMapper<TComponent> : IEntityPropertyMapper
...
+	public static class ComponentAttributesMapper
+	{
+		// 6.0 TODO: Move to IComponentAttributesMapper
+		public static void LazyGroup(this IComponentAttributesMapper mapper, string name)
+		{
+			if (mapper is ComponentMapper component)
+			{
+				component.LazyGroup(name);
+			}
+		}
```
Comment (fredericDelaporte) [inline src/NHibernate/Mapping/ByCode/IComponentMapper.cs]: I was wondering if it should throw when it is not a `ComponentMapper`. But it seems the pattern for not supported properties is to be just no-op, in mapping ByCode. So well, let it be no-op here too.

File: src/NHibernate/Persister/Entity/AbstractEntityPersister.cs [hunk 4]
```diff
@@ -4247,7 +4354,12 @@ public bool HasUpdateGeneratedProperties
 
 		public void AfterInitialize(object entity, bool lazyPropertiesAreUnfetched, ISessionImplementor session)
```
Comment (fredericDelaporte) [inline src/NHibernate/Persister/Entity/AbstractEntityPersister.cs]: It should be obsoleted.

File: src/NHibernate/Tuple/Entity/AbstractEntityTuplizer.cs [hunk 5]
```diff
@@ -228,6 +228,10 @@ public virtual void AfterInitialize(object entity, bool lazyPropertiesAreUnfetch
 		{
```
Comment (fredericDelaporte) [inline src/NHibernate/Tuple/Entity/AbstractEntityTuplizer.cs]: It should be obsoleted.

File: src/NHibernate/Tuple/Entity/PocoEntityTuplizer.cs [hunk 6]
```diff
@@ -214,13 +216,18 @@ protected virtual IProxyFactory BuildProxyFactoryInternal(PersistentClass @class
 		}
 
 		public override void AfterInitialize(object entity, bool lazyPropertiesAreUnfetched, ISessionImplementor session)
```
Comment (fredericDelaporte) [inline src/NHibernate/Tuple/Entity/PocoEntityTuplizer.cs]: It should be obsoleted. Or removed, by already calling the new method in the base class.

**Comment by hazzik** [comment]
Yes

**Comment by maca88** [comment]
I've also checked, thanks @fredericDelaporte.

**Pull request merged**